Eastern-Research-Group/expert-query
# Feature/149 reporting cycle field changes

## Related Issues:

- [EQ-149](https://jira.epa.gov/secure/RapidBoard.jspa?rapidView=865&view=detail&selectedIssue=EQ-149)

## Main Changes:

- Switched the `reportingCycle` field to a required, single-select input that
  defaults to "Latest". This option holds a `null` value, which the server then
  interprets as "use the most recent cycle".
- Separated `InputState` into a filter state, a source state (for source select
  dropdown values), and a single `format` `useState` variable (since this is a
  universal field.
- Removed the `format` parameter from the query hash parameters. It seemed
  unnecessary, since the hash query is used to populate the query, not control
  the entire UI state. The `format` parameter is still passed to the API.
- Restructured some TypeScript files. Shared types can be imported from
  `types/index.ts`, but a global augmentation is still present for the `Array`
  constructor. Module declarations for SVGs and the `glossary-panel` are in
  separate \*.d.ts files.
- Moved components to the top of each file, with the default export as the
  component the file is named after, and moved utilities, hooks, constants, and
  types to the bottom. I think this aids readability, but let me know if you
  have an opinion otherwise.

## Steps To Test:

1. Navigate to http://localhost:3000/attains/assessmentUnits.
2. The option selected by default in the _Reporting Cycle_ field should be "Latest".
3. Click the **Download** button, and take note of the number of rows returned.
4. In the _Reporting Cycle_ field, select the option just below "Latest" (i.e. the most recent year).
5. Click the **Download** button again, and verify that the row count matches the previous count.
6. Click "Continue" in the download modal, and confirm that the `reportingCycle` field has the value of the latest year for all rows (currently 2020).
7. Navigate to http://localhost:9090/attains/data/assessmentUnits?columns=reportingCycle.
8. Confirm that all rows returned have a value of 2020 for the `reportingCycle` field.
9. Navigate to http://localhost:9090/attains/data/assessmentUnits/count?columns=reportingCycle.
10. Confirm that the result count matches the counts noted previously.

## NOTE
I apologize in advance for the size of this PR. A certain amount of refactoring was necessary to avoid duplication when modifying the _Reporting Cycle_ input, and this led to a lot of restructuring.

The following commits are most relevant to the ticket at hand:
- f26dc7d5d1bdb98128191f53555e6f4fa21d63da
- 0751658c6c225d98bdc765bbee19430f766b093b
- 23e51d14af3e08730430eb7065ded26735283eb9
